Gelato scheduler automation

joeljesu opened the pull request:
Adding in a simple Gelato resolver contract to make it easier to automate Flow Scheduling with Gelato.

Resolver contract contains a `checker` function that returns the appropriate transaction payload (`executeCreateFlow`, `executeDeleteFlow`, or nothing) so Gelato can execute appropriately.

joeljesu commented:
Looking at information from @donoso_eth's [Gelato v2 x Superfluid guide](https://medium.com/coinmonks/gelato-v2-superfluid-eb13166ac414), this can be improved further before merging.

kasparkallas commented on packages/automation-contracts/scheduler/contracts/SchedulerGelatoResolver.sol:
```diff
@@ -0,0 +1,81 @@
+
+        }
+
+        else if ( flowSchedule.startDate != 0 &&
```
I feel like the condition for "create flow" should be before "delete flow".

joeljesu replied:
I was under the impression there could be a situation where flowSchedule.startDate + startMaxDelay is greater than flowSchedule.endDate and in that case, the resolver could return a start payload when it shoud be a delete payload if the condition for start came before delete

However, looking here: https://github.com/superfluid-finance/protocol-monorepo/blob/gelato-scheduler-automation/packages/automation-contracts/scheduler/contracts/FlowScheduler.sol#L158 - if the flow was started, then flowSchedule.startDate would be zeroed out and returning a start payload would be blocked by `flowSchedule.startDate != 0` on line 52. 

I can make this adjustment!

kasparkallas commented on packages/automation-contracts/scheduler/contracts/SchedulerGelatoResolver.sol:
```diff
@@ -0,0 +1,81 @@
+    ISuperToken
+} from "@superfluid-finance/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol";
+
+contract SchedulerGelatoResolver {
```
"FlowSchedulerGelatoResolver" or something more specific could be better?

ngmachado commented:
Already talked with @sunnyjaycer about this PR. 

## Recap:
- Check GelatoV2
- Rename canonical name to real name flowScheduler.
- Review tests and remove tests not related to gelato integration. _(if needed to add more tests do it in the proper test file)_
- Take into consideration @kasparkallas naming proposal.

joeljesu commented:
Keeping Gelato v2 out of scope here, will do some homework and incorporate support later

ngmachado commented:
> Looking at information from @donoso_eth's [Gelato v2 x Superfluid guide](https://medium.com/coinmonks/gelato-v2-superfluid-eb13166ac414), this can be improved further before merging.

It was you that brought this to scope 😬

going to finish my review tomorrow. Thank you for the fast changes!

ngmachado commented on packages/automation-contracts/scheduler/contracts/FlowSchedulerResolver.sol:
```diff
@@ -0,0 +1,121 @@
+// solhint-disable not-rely-on-time
+pragma solidity ^0.8.0;
+
+import {console} from "forge-std/console.sol";
```
remove

ngmachado commented on packages/automation-contracts/scheduler/old-resolver-test.txt:
```diff
@@ -0,0 +1,538 @@
+// SPDX-License-Identifier: UNLICENSED
```
is this file needed? if not delete it

ngmachado commented on packages/automation-contracts/scheduler/test/FlowSchedulerResolver.t.sol:
```diff
@@ -0,0 +1,489 @@
+
+    /// @dev Constants for Testing
+    uint32 internal defaultStartDate = uint32(block.timestamp + 1);
+    uint32 testNumber;
```
delete it

hellwolf commented:
@ngmachado any holdup for merging?

ngmachado commented:
@hellwolf for me is good to merge. We need another approve to merge it.

hellwolf commented:
![image](https://user-images.githubusercontent.com/186660/214021302-fffbd37b-fd20-4a59-8975-35340ce8a6d7.png)

@ngmachado you'd need to squash and force-push

ngmachado committed:
gelato resolver and tests

tweak comments and remove console logs

more specific tests, renamings, permission conditionals in resolver

cherry

ngmachado committed:
Merge branch 'dev' into gelato-scheduler-automation
